CS-458/academic-requirements
# AR-156 Update requirements processing

# Jira Ticket

Jira ticket number with link: [AR-156](https://academic-cs458.atlassian.net/jira/software/projects/AR/boards/1?isInsightsOpen=true&selectedIssue=AR-156)

# Description of Changes
_Please describe what this pull request should accomplish_
This updates requirements to count courses that are repeatable for credit more than once. It also includes a refactoring of a lot of the requirements code to make it easier to read and have fewer bugs.
**THIS INCLUDES DATABASE CHANGES**

# Type of Change

- [x] Feature Update

# How has this been tested?
_Please describe the tests that you used to verify your changes._

**Automated Tests (Preferred)**
- [x] Check Major/Concentration Requirements Processing
- [x] Check repeatable for credit courses
- [x] Check General Requirements Processing
- [x] Check ART/HUM or SBSCI Requirements Processing
- [x] Check ARNS Requirements Processing

**Manual Tests**
_Please explain why they are not automated._
At this point the only manual testing required would be to visually drag a course and see that the numbers update on the screen. I would like to automate this once the schedule page is able to be tested but I don't think we have it rendering quite yet.

# Checklist
- [x] My code follows ES Lint guidelines
- [x] My solution meets all of the acceptance criteria for the user story
- [x] My code contains necessary comments and documentation
- [x] I have added tests that are useful and cover the new changes
